satellite/satellitedb/containment.go

```diff
@@ -0,0 +1,116 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package satellitedb
+
+import (
+	"context"
+	"database/sql"
+
+	"storj.io/storj/pkg/audit"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	dbx "storj.io/storj/satellite/satellitedb/dbx"
+)
+
+type containment struct {
+	db *dbx.DB
+}
+
+// Get gets the pending audit by node id
+func (containment *containment) Get(ctx context.Context, id pb.NodeID) (*audit.PendingAudit, error) {
+	if id.IsZero() {
+		return nil, audit.ContainError.New("node ID empty")
+	}
+
+	pending, err := containment.db.Get_PendingAudits_By_NodeId(ctx, dbx.PendingAudits_NodeId(id.Bytes()))
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, audit.ErrContainedNotFound.New(id.String())
+		}
+		return nil, audit.ContainError.Wrap(err)
+	}
+
+	return convertDBPending(pending)
+}
+
+// IncrementPending creates a new pending audit entry, or increases its reverify count if it already exists
+func (containment *containment) IncrementPending(ctx context.Context, pendingAudit *audit.PendingAudit) error {
+	tx, err := containment.db.Open(ctx)
+	if err != nil {
+		return audit.ContainError.Wrap(err)
+	}
+
+	statement := containment.db.Rebind(
+		`INSERT INTO pending_audits (node_id, piece_id, stripe_index, share_size, expected_share_hash, reverify_count)
+		VALUES (?, ?, ?, ?, ?, ?)
+		ON CONFLICT(node_id)
+		DO UPDATE SET reverify_count = pending_audits.reverify_count + 1`,
+	)
+	_, err = tx.Tx.ExecContext(ctx, statement,
+		pendingAudit.NodeID.Bytes(), pendingAudit.PieceID.Bytes(), pendingAudit.StripeIndex, pendingAudit.ShareSize, pendingAudit.ExpectedShareHash, pendingAudit.ReverifyCount,
+	)
+	if err != nil {
+		return audit.ContainError.Wrap(err)
```

Comment:
We have to rollback the transaction here if there is an error.

The best way to have a single point for executing _commit_ or _rollback_ is to use a `defer` function call

Reply:
I was thinking of doing:
```
	return audit.ContainError.Wrap(errs.Combine(err, tx.Rollback()))
```
since I've seen it in a few other places. But how would you suggest using the defer?

Reply:
That's ok if it's fine to you, like me, to do the rollback on each error return since the `tx` creation.

I'm not a big fun of `defer` with logic which involve return named parameters, but it's worse to have duplicated logic because `tx.Rollback` can return an error, that's why I suggested, however using `errs` package it doesn't happen, so I prefer what you did.

Reply:
Now that I've seen the changes, I would say what someone could prefer to use a `defer`; it would be because the logic isn't very simple (there are 4 points where the `Rollback` must be executed) so it's easy to forget to call `Rollback` in one of them or if the function get refactored in the future, so a `defer` would keep the `Rollback` and `Commit` execution in one single point just after the `tx` creation.

NOTE I'm not asking for changing in that way, it has that advantage but `defer` with named return parameters doesn't also produce a 100% clear logic, so as usual there are advantages and drawbacks.

I mostly wanted to give you the different rationales, I'm fine with both.